lmdeploy/serve/async_engine.py

```diff
@@ -85,47 +150,32 @@ def __call__(self,
               1.0 means no penalty
             ignore_eos (bool): indicator for ignoring eos
-            do_preprocess (bool): whether pre-process the messages.
         """
         return self.batch_infer(prompts,
+                                gen_config=gen_config,
+                                chat_template_config=chat_template_config,
                                 request_output_len=request_output_len,
                                 top_k=top_k,
                                 top_p=top_p,
                                 temperature=temperature,
                                 repetition_penalty=repetition_penalty,
                                 ignore_eos=ignore_eos,
-                                do_preprocess=do_preprocess,
                                 **kwargs)
 
     def stop_session(self, session_id: int):
         """Stop a session by a session_id."""
-        input_ids = [self.tm_model.eos_id]
-        stop_generator = self.tm_model.create_instance()
-        for outputs in stop_generator.stream_infer(session_id,
-                                                   input_ids,
-                                                   request_output_len=0,
-                                                   sequence_start=False,
-                                                   sequence_end=False,
-                                                   stop=True):
-            pass
-        if str(session_id) in self.id2generator and self.id2generator[str(
-                session_id)] not in self.gens_set:
-            self.gens_set.add(self.id2generator[str(session_id)])
+        if str(session_id) in self.id2generator:
+            self.id2generator[str(session_id)].cancel(session_id)
+            if self.id2generator[str(session_id)] not in self.gens_set:
+                self.gens_set.add(self.id2generator[str(session_id)])
 
     def end_session(self, session_id: int):
         """Clear a session by a session_id."""
-        input_ids = [self.tm_model.eos_id]
-        end_generator = self.tm_model.create_instance()
-        for outputs in end_generator.stream_infer(session_id,
-                                                  input_ids,
-                                                  request_output_len=0,
-                                                  sequence_start=False,
-                                                  sequence_end=True):
-            pass
-        self.id2step[str(session_id)] = 0
-        if str(session_id) in self.id2generator and self.id2generator[str(
-                session_id)] not in self.gens_set:
-            self.gens_set.add(self.id2generator[str(session_id)])
+        if str(session_id) in self.id2generator:
+            self.id2generator[str(session_id)].end(session_id)
+            self.id2step.pop(str(session_id))
+            if self.id2generator[str(session_id)] not in self.gens_set:
+                self.gens_set.add(self.id2generator[str(session_id)])
```

Comment:
is `self.gens_set` a `set` object? It is legal to add repeated objects into a set. you can just remove the `if` above.